### ISSUE TITLE
🦴 test(conftest): skeleton — shared fixture seam

### ISSUE THREAD

**Issue by CedrickArmel** [CedrickArmel/radiologist#98]
# 🦴 Skeleton: shared fixture seam

## Context

This is the skeleton issue for the **Classist outside-in TDD compliance + pytest fixture
consolidation** epic. It establishes the shared fixture seam that every slice issue depends on,
so the five slices (#99–#103) can proceed in parallel afterwards. **All other issues in this epic
block on this one — it must land first.**

This issue is test-infrastructure only. It introduces no product behavior. It moves the
`sys.path` shim to a single repo-root `conftest.py`, consolidates value/transform fixtures and
ONNX builders into canonical locations, and adds new fixtures (`dm`, `ckpt_path`, `lmodule`) plus
the predictor/nonzero fixtures the slices rely on. After this issue, the existing test suite must
still be green — no test is rewritten here; only fixtures and shim location change, kept
behavior-compatible.

### `conftest.py` — repo root — **NEW**

Single session-global conftest collected once for all `testpaths`. Inserts the `src/` directory
of **all five** packages onto `sys.path` so `radiologist.*` is importable everywhere. Replaces the
five per-package shim lines.

```python
import sys
from pathlib import Path

_ROOT = Path(__file__).parent
for _pkg in (
    "radiologist-utils",
    "radiologist-etl",
    "radiologist-core",
    "radiologist-inference",
    "radiologist-registry",
):
    sys.path.insert(0, str(_ROOT / _pkg / "src"))
```

> The repository enforces a license header on all source files via a pre-commit hook. Do **not**
> add the header by hand; let the hook insert it.

### `radiologist-utils/tests/conftest.py` — **EDIT**

Remove the `sys.path.insert(0, .../src)` line and its now-unused `import sys` / `from pathlib import
Path` if they are used for nothing else. Keep all other fixtures.

### `radiologist-etl/tests/conftest.py` — **EDIT**

Remove the shim line. This file also imports `numpy`, `pytest`, `PIL.Image` for other fixtures —
keep those; only the `sys.path`/`Path` shim goes.

### `radiologist-registry/tests/conftest.py` — **EDIT**

Remove the shim line and its now-unused imports.

### `radiologist-inference/tests/conftest.py` — **EDIT**

Remove the shim line. Then extend the canonical ONNX layer (which already exists here):

- Add a `feat_nonzero: bool = False` parameter to `build_det_onnx`. When `True`, the
  `feature_maps` output is filled with a deterministic non-zero constant instead of zeros, so
  Score-CAM has gradient signal to attribute. Default `False` preserves current behavior.
- Add fixture `det_onnx_path_nonzero` returning `build_det_onnx(tmp_path, feat_nonzero=True)`.
- Add fixtures `predictor_with_mcd`, `predictor_without_mcd` building real `Predictor` instances
  from real ONNX files (det + mcd for the former, det only for the latter).
- Add fixture `sample_image` returning a small deterministic `np.ndarray` (or PNG-bytes / file as
  the predictor's public input contract requires — match what `test_mc_dropout.py` currently feeds).

Keep the canonical builder names `build_det_onnx` / `build_mcd_onnx` and `_add_metadata` exactly so
`from conftest import build_det_onnx, build_mcd_onnx` (used by `test_app.py`) keeps resolving.

### `radiologist-core/tests/conftest.py` — **EDIT**

- Remove the shim line (the root conftest now owns it). Keep all other content.
- Make `label_map`, `classes`, `batch_size` **session-scoped** (`@pytest.fixture(scope="session")`).
- Replace `train_transform` and `eval_transform` with a single **session-scoped** `transform`
  fixture (the merged transform). Delete the two old fixtures.
- Add a function-scoped composed `dm` fixture that builds a ready `WebDatasetDataModule` from the
  existing fixtures (`split_manifest_uri`, `transform`, `label_map`/`classes`, loader partials,
  `batch_size`). It must be function-scoped because it transitively depends on `tmp_path`.
- Add a function-scoped `ckpt_path` fixture producing a **real Lightning checkpoint** loadable by
  `LModule.load_from_checkpoint` (see contract below).
- Add a function-scoped `lmodule` fixture returning a real tiny `LModule` instance. This is the
  shared `LightningModule` used by callback/module tests that run a real `pl.Trainer`. Having it
  in conftest avoids repeating the tiny-net construction in every test file.

### `pyproject.toml` — repo root — **VERIFY ONLY (no change expected)**

`[tool.pytest.ini_options]` already declares `testpaths` (all five `tests` dirs) and
`addopts = "--import-mode=importlib --cov=radiologist --cov-report=term-missing"`. Confirm it is
present and unchanged. Only edit if the root conftest is not being collected — it should not
require changes.

### `radiologist-registry/src/radiologist/registry/__init__.py` — **VERIFY ONLY (no change expected)**

`WandbRegistry` is already exported in `__all__`. Confirm and move on.

### `radiologist-core/tests/conftest.py`

```python
@pytest.fixture(scope="session")
def label_map() -> dict: ...
    # contract: canonical class→index map; stable across the session, no tmp_path dependency

@pytest.fixture(scope="session")
def classes() -> list: ...
    # contract: ordered class names matching label_map

@pytest.fixture(scope="session")
def batch_size() -> int: ...
    # contract: the batch size used by the loader partials

@pytest.fixture(scope="session")
def transform(): ...
    # contract: single torchvision transform pipeline used for both train and eval;
    #           replaces the former train_transform / eval_transform pair

@pytest.fixture()
def dm(split_manifest_uri, transform, label_map, classes, batch_size, ...):
    # contract: returns a constructed WebDatasetDataModule wired from the shared fixtures,
    #           with `transform` passed as BOTH train_transform and eval_transform;
    #           function-scoped (depends transitively on tmp_path)

@pytest.fixture()
def lmodule() -> "LModule":
    # contract: returns a real LModule wrapping a tiny conv net (same architecture as the
    #           ckpt_path fixture uses). Suitable for passing to pl.Trainer.fit().
    # function-scoped; creating it is cheap.

@pytest.fixture()
def ckpt_path(tmp_path) -> str:
    # contract: writes a real Lightning checkpoint to tmp_path and returns its path;
    #           LModule.load_from_checkpoint(ckpt_path) succeeds and yields a usable module
```

`lmodule` reference implementation:

```python
@pytest.fixture()
def lmodule():
    from functools import partial
    import torch.nn as nn
    from torchmetrics.classification import MulticlassFBetaScore
    from radiologist.core import FocalLoss, LModule

    net = nn.Sequential(
        nn.Conv2d(3, 4, 3, padding=1), nn.ReLU(), nn.Dropout(p=0.5),
        nn.AdaptiveAvgPool2d((1, 1)), nn.Flatten(), nn.Linear(4, 2),
    )
    return LModule(
        net=net, loss=FocalLoss(),
        metric=partial(MulticlassFBetaScore, beta=1.0, num_classes=2),
        optimizer=partial(torch.optim.Adam, lr=1e-3),
    )
```

`ckpt_path` reference implementation:

```python
@pytest.fixture()
def ckpt_path(tmp_path, lmodule):
    import pytorch_lightning as pl
    import torch

    path = str(tmp_path / "test.ckpt")
    ckpt = {
        "epoch": 0, "global_step": 0,
        "pytorch-lightning_version": pl.__version__,
        "state_dict": lmodule.state_dict(),
        "hyper_parameters": {},
    }
    torch.save(ckpt, path)
    return path
```

### `radiologist-inference/tests/conftest.py`

```python
def build_det_onnx(
    tmp_path,
    priors=None,         # Optional[dict]
    filename: str = "model_det.onnx",
    feat_nonzero: bool = False,
) -> str:
    # contract: deterministic 2-class ONNX classifier; logits via Softmax, feature_maps (1,64,7,7);
    #           feat_nonzero=True fills feature_maps with a deterministic non-zero constant for
    #           Score-CAM attribution; priors embeds training_prior metadata when given

def build_mcd_onnx(tmp_path, filename: str = "model_mcd.onnx") -> str:
    # contract: stochastic RandomUniform->Softmax ONNX model (unchanged)

@pytest.fixture()
def det_onnx_path(tmp_path) -> str: ...          # build_det_onnx(tmp_path)
@pytest.fixture()
def det_onnx_path_nonzero(tmp_path) -> str: ...  # build_det_onnx(tmp_path, feat_nonzero=True)
@pytest.fixture()
def mcd_onnx_path(tmp_path) -> str: ...          # build_mcd_onnx(tmp_path)
@pytest.fixture()
def predictor_with_mcd(tmp_path): ...            # real Predictor from real det + mcd ONNX
@pytest.fixture()
def predictor_without_mcd(tmp_path): ...         # real Predictor from real det ONNX only
@pytest.fixture()
def sample_image(): ...                           # deterministic np.ndarray input
```

### `conftest.py` (root)

```python
# contract: on collection, all five radiologist.* namespaces are importable;
#           no leaf tests/conftest.py performs a sys.path.insert
```

## Acceptance criteria

- [ ] `radiologist.utils`, `radiologist.etl`, `radiologist.core`, `radiologist.inference`,
      `radiologist.registry` all import in tests with no leaf-package `sys.path` shim present.
- [ ] `make test` (all five packages) is green with the new fixtures in place and no existing test
      rewritten.
- [ ] `LModule.load_from_checkpoint(<value of ckpt_path>)` returns a usable module without patching.
- [ ] Requesting `dm` yields a `WebDatasetDataModule` whose splits load from the real JSONL provided
      by `split_manifest_uri`.
- [ ] Requesting `lmodule` yields a real `LModule` that `pl.Trainer(fast_dev_run=True).fit()` accepts.
- [ ] Requesting `det_onnx_path_nonzero` yields an ONNX model whose feature maps are non-zero.
- [ ] `from radiologist.registry import WandbRegistry` succeeds.
- [ ] mypy clean; pytest green.

## Out of scope

- Rewriting any test body to remove a mock — that is the work of slices #99–#103. This issue only
  provides the fixtures those slices consume, and keeps the current suite green.

## Technical notes

- `dm`, `ckpt_path`, and `lmodule` must be **function-scoped**: they depend transitively on
  `tmp_path` or create stateful objects.
- `ckpt_path` depends on `lmodule`; reuse the fixture rather than constructing the tiny net twice.
- The `from conftest import build_det_onnx, build_mcd_onnx` import in `test_app.py` works only
  because importlib mode makes the `tests/` dir importable — keep the canonical builder names.
- Do not insert a license header manually; the pre-commit hook handles it.
- Use `Optional[dict]` from `typing` for annotations in conftest rather than `dict | None` to stay
  3.10-compatible without `from __future__ import annotations`.
